src/commands/multi_block/monitor.rs

```diff
@@ -237,93 +421,145 @@ where
+            prometheus::observe_mined_solution_duration(mining_duration.as_millis() as f64);
             sol
         }
-        (Err(e), _) => {
+        Err(e) => {
             return Err(e);
         }
     };
+    // Capture score before potentially moving paged_raw_solution
+    let submitted_score = paged_raw_solution.score;
 
-    let storage_head = utils::storage_at_head(&client, listen).await?;
-
-    match get_submission(&storage_head, round, signer.account_id(), n_pages).await? {
+    // Check again if we have submitted something while mining
+    let storage_after_mining = storage_at_finalized(&client).await?;
+    match get_submission(&storage_after_mining, round, signer.account_id(), n_pages).await? {
         CurrentSubmission::Done(score) => {
             // We have already submitted the solution with a better score or equal score
-            if !score_passes_strategy(paged_raw_solution.score, score, submission_strategy) {
+            if !score_passes_strategy(submitted_score, score, submission_strategy) {
+                log::info!(target: LOG_TARGET, "Mined solution score {:?} not better than already submitted score {:?} for round {}. Skipping.", submitted_score, score, round);
+                // Ensure it's recorded locally if somehow missed before
+                if !submitted_rounds.contains(round).await {
+                    submitted_rounds.insert(round, n_pages, score, None).await;
+                }
                 return Ok(());
             }
             // Revert previous submission and submit the new one.
-            bail(listen, &client, signer.clone()).await?;
+            log::warn!(target: LOG_TARGET, "Mined solution score {:?} is better than submitted score {:?} for round {}. Bailing previous submission.", submitted_score, score, round);
+            bail(Listen::Finalized, &client, signer.clone()).await?;
+            // Clear local tracking as we are bailing
+            submitted_rounds
+                .remove(round, Some(RoundUntrackingEvent::Bailed))
+                .await;
         }
         CurrentSubmission::Incomplete(s) => {
-            // Submit the missing pages.
-            if s.score() == paged_raw_solution.score {
+            // Submit the missing pages if score matches.
+            if s.score() == submitted_score {
+                log::info!(target: LOG_TARGET, "Found incomplete submission for round {} with matching score {:?}. Submitting missing pages.", round, submitted_score);
                 let mut missing_pages = Vec::new();
-
                 for page in s.get_missing_pages() {
-                    let solution = paged_raw_solution.solution_pages[page as usize].clone();
-                    missing_pages.push((page, solution));
+                    // Clone only needed pages
+                    if let Some(solution_page) =
+                        paged_raw_solution.solution_pages.get(page as usize)
```

Comment:
this should be infallible, page idx must be valid I would be okay with an unwrap/expect here but fair enough